![Culebront Logo](./SupplementaryFiles/Culebront-LOGO2.png)

## Table of Contents
<!-- TOC depthFrom:2 depthTo:3 withLinks:1 updateOnSave:1 orderedList:0 -->
- [About this package](#about-this-package)
- [Installation](#installation)
- [Rules in CulebrONT](#rules)
- [Running CulebrONT](#running-culebront)
- [Parameters](#Parameters)
- [Singularity images](#singularity)
- [Cluster_config.yaml](#prepclusteryaml)
- [Launching CulebrONT on HPC clusters](#sbatch)
- [Output on CulebrONT](#output)
- [Citation](#citation)
- [Useful notes](#notes)
- [License](#license)
<!-- /TOC -->

<a name="about-this-package"></a>
## About this package

Using data from long reads obtained by Oxford Nanopore sequencing Technologies makes genome assembly easier, in particular to solve repeats and structural variants, in prokaryotic as well as in eukaryotic genomes, resulting in increased contiguity and accuracy. Plenty of softwares and tools are released or updated every weeks, and a lot of species see their genome assembled using those tools. That’s right, but which assembly tool could give the best results for my favorite organism? CulebrONT can help you! CulebrONT is a scalar, modulable and traceable snakemake pipeline.

<a name="installation"></a> 
## Installation

CulebrONT uses python >= 3.7 and Snakemake >= 5.10.0
  
``` bash
# On i-Trop HPC
module load system/singularity/3.3.0
module load system/python/3.7.2

# On IFB HPC
module load singularity 
module load python/3.7
module load graphviz/2.40.
```

* Install or update CulebrONT

``` bash
git clone https://github.com/SouthGreenPlatform/culebront_pipeline.git
```

<a name="rules"></a>
## Rules in Culebront

Assembly, circularisation, polishing and correction steps are included in CulebrONT, and can be activated (or not) according to user’s requests. The most relevant tools commonly used for each step are integrated, as well as a lot of quality control tools. CulebrONT also generates a report compiling information obtained in every step. CulebrONT is an open source solution to help you making decisions for your own assembly.

### From assembly to correction

CulebrONT is really flexible to assembly and circularise (or not) assembled molecules, polishing and correct assemblies. 

For assemblies, you can launch at least one of assemblers included on culebrONT and pipe assembly with circularisation, polishing and correction steps as well as quality pipeline. 

For circularisation, you can activate/deactivate circularisation steps if need.
If you are interested on eukaryotic organims and circularisation is not necessary use CIRCULAR=False on *config.yaml* file.

Dags here are showing differences between deactivated (left) and activated (right) CIRCULAR step on configuration file. 

![global](images/assembly_to_correction.png)

##### Assembly 
CulebrONT includes (at the moment) three usefully assemblers : Flye, Miniasm and Canu. 

Included tools :
* Flye version >= 2.6 https://github.com/fenderglass/Flye
* Canu version >= 1.9 https://canu.readthedocs.io/en/latest/quick-start.html
* Miniasm version >= 0.3 https://github.com/lh3/miniasm + Minipolish version >= 0.1.2 https://github.com/rrwick/Minipolish

##### Circularisation

If a assembled molecule is circular (CIRCULAR=True), this molecule is tagged and will be treated specially in pipeline. We implemented taggind and rotation of circular molecule before each racon polishing step and we fixing start position on circular genome. This could be good when multiple alignments are envisaged.  

* Circularisation (CIRCULAR=True) step is running activating *--plasmids* option on Flye. 
* *Ciclator* is used to circularise assembly from Canu. Circlator will attempt to identify each circular sequence and output a linearised version of it. 
* Circularisation to Miniasm is already insured by minipolish.

Included tools :
* Circlator version >= 1.5.5 https://github.com/sanger-pathogens/circlator

##### Polishing
Polishing step is insured by Racon. Racon correct raw contigs generated by rapid assembly methods. Choose how many rounds of racon you would run (constrains from 1 to 9 rounds). CulebrONT will recursively do it for you. 

Included tools :
* Racon version >= 1.4.3 https://github.com/isovic/racon

##### Correction 
Correction can improved consensus sequence for a draft genome assembly. We have included Nanopolish and Medaka on correction steps. With CulebrONT you can now training a model Medaka and use it directly to obtain a consensus from you favorite organism. 

Included tools :
* Medaka Medaka-gpu version >= 0.11.5 https://github.com/nanoporetech/medaka
* Nanopolish version >= 0.11.3 https://nanopolish.readthedocs.io/en/latest/index.html#

### Checking assembly quality 

A variety of useful tools exist for check high accuracy assembly.

![global](images/quality.png)

##### Quality tools
 Using CulebrONT, BUSCO and QUAST are launched by default. BUSCO could help to check if we have a good assembly by searching the expected single-copy ortholog in any newly-sequenced genome from a appropriate phylogenetic clade. QUAST is a good starting point to help evaluate the quality of assemblies. It provides many helpful contiguity statistics. BUSCO and QUAST results are summary on report.
* BUSCO and QUAST statistics will be calculated by the QUALITY (ASSEMBLY, POLISHING, CORRECTION) activated steps. You must to activate at least one over three QUALITY options on *config.yaml* file.

Included tools :
* Busco version >= 4.0.5
* Quast version >= 5.0.2

##### Optional Quality tools

Check also quality assembly by using Bloobtools or/and Assemblytics or/and KAT. Use Weesam to check coverage of reads over you assembly (only for small genome size). Align assemblies from several assembles (or steps : assembly, polishing, correction) and compare it using Mauve.

Included tools :  
* Bloobtools version >= 1.1.1
* Assemblytics version >= 1.2
* KAT version >= 2.4.1
* Weesam version > 1.6
* Mauve > 2.4.0.snapshot_2015_02_13

##### Report 
CulebrONT generates a report summary statistics from different steps of the pipeline.

Report needs preinstalled `devtools::install_github("strengejacke/strengejacke")` R dependance and also 'plotly', 'dplyr', 'optparse', 'htmltools', 'rmdformats', 'magrittr', 'yaml', 'png', 'here', 'htmlwidgets'.

Don't worry !! The whole of packages used to report are available on  the *R.def* singularity image that you can use from singularity hub or build from Containers repository https://github.com/SouthGreenPlatform/culebront_pipeline/tree/master/Containers if you have sudo superpowers.

<a name="running-culebront"></a>
## Running CulebrONT

### Prepare config.yaml

To run the pipeline you have to give data paths and activate/deactivate options in every step from config.yaml file.

##### 1. Give data 

First, give path from data in configuration file

```yaml
DATA:
    FASTQ: '/path/to/repertory/with/fastq/'
    REF: '/path/to/referencefile.fasta'
    GENOME_SIZE: '1m'
    FAST5: '/path/to/repertory/with/fast5/'
    ILLUMINA: '/path/to/repertory/with/illumina/'
    OUTPUT: '/path/to/repertory/output/'
    CIRCULAR : True/False
```

* FASTQ: CulebrONT takes as input a fastq repertory ('FASTQ'). Every fastq file would contain the whole of reads to be assembled. Each fastq file found in this repertory will be assembled independently. Fastq files can be compressed or not. Naming convention accepted by CulebrONT is *NAME.fastq.gz* or *NAME.fq.gz* or *NAME.fastq* or *NAME.fq*.

* REF: Only a reference will be used by CulebrONT. Put the reference path in the 'REF' DATA parameter.

* GENOME_SIZE : Genome size (m,g,k) of reference.

* FAST5: Nanopolish needs 'FAST5' files to training steps. Please give the path of fast5 repertory in 'FAST5' DATA parameter. Inside this 'fast5' repertory, a subrepertory with same name of fastq (before.fastq.gz) is necessary. For example : if in the FASTQ repertory we have run1.fastq.gz and run2.fastq.gz culebrONT waits fast5/run1 and fast5/run2 repertories in 'FAST5' path.   

* ILLUMINA : put path to repertory with illumina fastq.gz to KAT quality. Preferentially paired end data.

* OUTPUT: output path repertory

* CIRCULAR : True or False to activate/deactivate circularisation steps (only to procaryote)


##### 2. Chose assemblers, polisher and correctors

Activate/deactivate assemblers and correctors. By default RACON is launched as POLISH tool after each activated assembly step. You must to activate at least a assembler and a corrector.

```yaml
ASSEMBLY:
    CANU : False
    FLYE : True
    MINIASM : False
CORRECTION:
    NANOPOLISH : True
    MEDAKA : False
```

##### 3. Chose quality tools

Activate/deactivate quality tools. By default BUSCO AND QUAST are launched if 'True' in QUALITY (ASSEMBLY OR/ET POLISHING OR/ET CORRECTION) steps.
 
 You must to activate at least a QUALITY step.

```yaml
QUALITY:
    ASSEMBLY : True
    POLISHING : True
    CORRECTION : True
```
Others quality tools are lauched only in *last assemblies* (BLOBTOOLS, ASSEMBLYTICS, WEESAM and KAT). 

KAT quality tool can be activate but Illumina reads are required.

```yaml
#### Others quality tools
    WEESAM: True
    BLOBTOOLS: True
    ASSEMBLYTICS: True
    KAT: True
```

Alignment of various assemblies derived from a fastq file for small genomes (<10-20Mbp) is also possible by using mauve. This one is 
lauched on the whole of quality steps activated by sample (fastq file).
A Fixstart step is possible before MSA to improve alignment on circular molecules.
* Fixstart will be deactivated if CIRCULAR is False
* Only activate MAUVE if you have more that 1 sample and more that 1 quality step.

```yaml
MSA:
    FIXSTART: True
    MAUVE: True
```

<a name="parameters"></a>
##### 4. Parameters in some specifics tools 

Specifically to Racon:
* Racon can be launch recursively from 1 to 9 rounds. 

Specifically to Medaka :
* If 'MEDAKA_TRAIN_WITH_REF' is activated, medaka launchs training using reference found in 'DATA/REF' param. Medaka does not take in count other medaka model parameters.
* If 'MEDAKA_TRAIN_WITH_REF' is deactivated, medaka does not launch training but use the model given in 'MEDAKA_MODEL_PATH'. If 'MEDAKA_MODEL_PATH' is empty, this param is forgotten and by default model *E.coli* is used.

```yaml
############ PARAMS ################
params:
    MINIMAP2:
        PRESET_OPTION: 'map-pb' # -x minimap2 preset option is map-pb by default (map-pb, map-ont etc)
    CANU:
        MAX_MEMORY: '15G'
        OPTIONS: '-fast'
    CIRCLATOR:
        OPTIONS: ''
    RACON:
        RACON_ROUNDS: 2 #1 to 9
    NANOPOLISH:
        # segment length to split assembly and correct it  default=50000
        NANOPOLISH_SEGMENT_LEN: '50000'
        # overlap length between segments  default=200
        NANOPOLISH_OVERLAP_LEN: '200'
        OPTIONS: ''
    MEDAKA:
        # if 'MEDAKA_TRAIN_WITH_REF' is True, medaka launchs training using reference found in DATA REF param. Medaka does not take in count other medaka model parameters below.
        MEDAKA_TRAIN_WITH_REF: True
        MEDAKA_MODEL_PATH: 'medakamodel/r941_min_high_g303_model.hdf5' # if empty this param is forgotten.
    BUSCO:
        DATABASE : 'Data-Xoo-sub/bacteria_odb10'
        MODEL : 'genome'
#        'SP' : 'caenorhabditis'
        SP : ''
    QUAST:
        REF: 'Data-Xoo-sub/ref/BAI3_Sanger.fsa'
        GFF: ''
        GENOME_SIZE_PB: 48000000
        #GENOME_SIZE_PB: 1000000
        OPTIONS : ''
    DIAMOND:
        DATABASE: 'Data-Xoo-sub/testBacteria.dmnd'
    MUMMER:
#         -l default 20
        MINMATCH : 100
#         -c default 65
        MINCLUSTER: 500
    ASSEMBLYTICS:
        UNIQUE_ANCHOR_LEN: 10000
        MIN_VARIANT_SIZE: 50
        MAX_VARIANT_SIZE: 10000
```

<a name="singularity"></a>
#### Singularity containers

Give to CulebrONT the build singularity containers path from your computer or cluster.  

Here singularity images found  on i-Trop HPC from SouthGreen platform.

```yaml
# cluster with scratch temporal repertory
SCRATCH = False
## @ITROP PATH
tools:
## ASSEMBLERS:
    CANU_SIMG : '/data3/projects/containers/CULEBRONT/canu-1.9.simg'
    FLYE_SIMG : '/data3/projects/containers/CULEBRONT/Flye-2.6.simg'
    MINIASM_SIMG : '/data3/projects/containers/CULEBRONT/miniasm-0.3.simg'
    MINIPOLISH_SIMG : '/data3/projects/containers/3.3.0/Minipolish-0.1.2.simg'
## CIRCULARISATION
    CIRCLATOR_SIMG : '/data3/projects/containers/CULEBRONT/Circlator-1.5.5.simg'
## POLISHERS:
    RACON_SIMG : '/data3/projects/containers/CULEBRONT/racon-1.4.3.simg'
## CORRECTION
    NANOPOLISH_SIMG : '/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg'
    MEDAKA_SIMG : '/data3/projects/containers/3.3.0/medaka-0.11.5.simg'
## QUALITY
    BUSCO_SIMG : '/data3/projects/containers/CULEBRONT/busco-4.0.5.simg'
    QUAST_SIMG : '/data3/projects/containers/CULEBRONT/quast-5.0.2.simg'
    WEESAM_SIMG : '/data3/projects/containers/CULEBRONT/weesam.simg'
    BLOBTOOLS_SIMG : '/data3/projects/containers/CULEBRONT/bloobtools-v1.1.1.simg'
    MINIMAP2_SIMG: '/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg'
    DIAMOND_SIMG : '/data3/projects/containers/CULEBRONT/Diamond-0.9.30.simg'
    MUMMER_SIMG : '/data3/projects/containers/CULEBRONT/mummer-4beta.simg'
    ASSEMBLYTICS_SIMG : '/data3/projects/containers/CULEBRONT/Assemblytics-1.2.simg'
    SAMTOOLS_SIMG : '/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg'
    KAT_SIMG : '/data3/projects/containers/CULEBRONT/kat-2.4.2.simg'
    MINICONDA_SIMG : 'shub://vibaotram/singularity-container:cpu-guppy3.4-conda-api'
    R_SIMG: '/data3/projects/containers/CULEBRONT/R.simg'
```

If you want to recovery singularity images from the singularity Hub, please use these paths :

```yaml
# cluster with scratch temporal repertory
SCRATCH : False

tools:
###### ASSEMBLERS:
      CANU_SIMG: 'shub://SouthGreenPlatform/CulebrONT:canu-1.9.def'
      FLYE_SIMG: 'shub://SouthGreenPlatform/CulebrONT:flye-2.6.def'
      MINIASM_SIMG : 'shub://SouthGreenPlatform/CulebrONT:miniasm-0.3.def'
      MINIPOLISH_SIMG : 'shub://SouthGreenPlatform/CulebrONT:Minipolish-0.1.2.simg' *
###### CIRCULARISATION
      CIRCLATOR_SIMG : 'shub://SouthGreenPlatform/CulebrONT:circlator-1.5.5.def'
###### POLISHERS:
      RACON_SIMG : 'shub://SouthGreenPlatform/CulebrONT:racon-1.4.3.def'
###### CORRECTION
      NANOPOLISH_SIMG : 'shub://SouthGreenPlatform/CulebrONT:nanopolish-0.11.3.def'
      MEDAKA_SIMG : 'shub://SouthGreenPlatform/CulebrONT:medaka-cpu-0.11.5.def'
###### QUALITY
      BUSCO_SIMG : 'shub://SouthGreenPlatform/CulebrONT:busco-4.def'
      QUAST_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      WEESAM_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      BLOBTOOLS_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      MINIMAP2_SIMG: 'shub://SouthGreenPlatform/CulebrONT:nanopolish-0.11.3.simg' *
      DIAMOND_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      MUMMER_SIMG : 'shub://SouthGreenPlatform/CulebrONT:mummer-4beta.def'
      ASSEMBLYTICS_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      SAMTOOLS_SIMG : 'shub://SouthGreenPlatform/CulebrONT:nanopolish-0.11.3.simg'
      KAT_SIMG : 'shub://SouthGreenPlatform/CulebrONT:quality.def'
      MINICONDA_SIMG : 'shub://vibaotram/singularity-container:cpu-guppy3.4-conda-api'
      R_SIMG : 'shub://SouthGreenPlatform/CulebrONT:R.simg'
```

Available recipes from containers are available in Containers repertory and also in https://github.com/SouthGreenPlatform/CulebrONT.git. Feel free of build them in your computer (or cluster) but you need root rights.

### Launching CulebrONT locally

<a name="Local"></a>

To launch CulebrONT on locally you should use the parameters `--use-singularity` and `--use-conda`. 
See the example bellow: 
```
snakemake -s Snakefile --configfile config.yaml --use-singularity  --use-conda --latency-wait 120
```

<a name="sbatch"></a>
### Launching CulebrONT on HPC clusters

<a name="prepclusteryaml"></a>
#### Preparing cluster configuration using cluster_config.yaml
On `cluster_config.yaml` you can add partition, memory and threads used by default for every rule. If more memory or threads are necessary, please adapt content of this file before running on cluster for every rule. For example give more memory to Canu and Medaka. 
 Here a example of configuration file used on i-Trop HPC.
 
```yaml
__default__:
    cpus-per-task : 4
    ntasks : 1
    mem-per-cpu : '2'
    partition : "normal"
    output : 'logs/stdout/{rule}/{wildcards}'
    error : 'logs/error/{rule}/{wildcards}'
#
#run_nanopolish :
#    cpus-per-task : 8
#    mem-per-cpu : '4'
#    partition : "long"
#
run_canu:
    cpus-per-task : 8
    mem-per-cpu : '4'
    partition : "long"

```

#### Profiles
In order to run CulebrONT on HPC cluster please follow directives found on https://github.com/Snakemake-Profiles/.

Here, we explain how to profile on slurm scheduler by using https://github.com/Snakemake-Profiles/slurm procedure.

```
$ mkdir -p ~/.config/snakemake
$ cd ~/.config/snakemake
$ pip install --user cookiecutter
$ cookiecutter https://github.com/Snakemake-Profiles/slurm.git
# Answer to question as well :
profile_name [slurm]: slurm-culebrONT
sbatch_defaults []: 
cluster_config []: cluster_config.yaml
Select advanced_argument_conversion:
1 - no
2 - yes
Choose from 1, 2 [1]: 2

$ cp /shared/home/$USER/culebront_pipeline/cluster_config.yaml .
# edit slurm-submit.py and upgrade this dictionay to slurm options

RESOURCE_MAPPING = {
	"cpus-per-task": ("cpus-per-task", "cpu"),
	"partition": ("queue", "partition"),
	"time": ("time", "runtime", "walltime"),
	"mem": ("mem", "mem_mb", "ram", "memory"),
	"mem-per-cpu": ("mem-per-cpu", "mem_per_cpu", "mem_per_thread"),
	"nodes": ("nodes", "nnodes"),
	"error": ("logerror", "e"),
	"output": ("logout", "o")}
```
SLURM Profile *slurm-culebrONT* is now created on : `/shared/home/$USER/.config/snakemake/slurm-culebrONT` repertory

##### submit_culebront.sh 

This is a typical launcher for using CulebrONT on a slurm cluster. You have to adapt it for configuration of you favorite cluster.
 
```
#!/bin/bash
#SBATCH --job-name culebrONT
#SBATCH --output slurm-%x_%j.log
#SBATCH --error slurm-%x_%j.log

module load system/singularity/3.3.0
module load system/python/3.7.2

snakemake --unlock

snakemake --nolock --use-singularity --use-conda --cores -p --verbose -s Snakefile \ 
--configfile config.yaml --latency-wait 60 --keep-going --restart-times 2 \ 
--rerun-incomplete --profile slurm-culebrONT \
```

This launcher can be launched on the cluster with:
```
sbatch submit_culebront.sh 
```

<a name="output"></a>
## Output on CulebrONT

Architecture of culebrONT output is showed above :
```
output_example_circ/
├── SAMPLE-1
│   ├── CANU
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   ├── FLYE
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   ├── MAUVE_ALIGN
│   ├── MINIASM
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   └── QUAST
│       ├── data
│       └── quast_results
├── LOGS
...
└── REPORT
    └── SAMPLE-1
```
* Similar Architecture observed by sample (fastq = SAMPLE-1 in example) is conserved for LOG files.

_Important_: To visualise the report created by culebrONT, transfer the whole of the repertory REPORT on your local machine before to open *report.html* on a navigator.

<a name="citation"></a>
## Citation

@Authors:

Aurore Comte (IRD) and Julie Orjuela (IRD) develop culebrONT.

Sebastien Ravel (CIRAD), Florian Charriat (CIRAD) help to snakemake bugs and tricks.

François Sabot (IRD) for implemented coverage tools.

Ndomassi Tando for helping on singularity containers.

Sebastien Cunnac (IRD) and Tram Vi (IRD) for test on real data and circularisation improvements.

<a name="notes"></a>
## Usefull notes

Previous to CulebrONT you could base-calling of arbitrarily multiplexed libraries across several Minion runs with sequencing quality control and gathers output files by genome for subsequent steps. For that use https://github.com/vibaotram/baseDmux.

#### Thanks

The authors acknowledge the IRD i-Trop HPC (South Green Platform) at IRD montpellier 
    for providing HPC resources that have contributed to this work.
   https://bioinfo.ird.fr/- http://www.southgreen.fr

<a name="licence"></a>
## License
Licencied under CeCill-C (http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html) and GPLv3
Intellectual property belongs to IRD.

