

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Table of Contents &mdash; CulebrONT 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/sphinx-argparse.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to culebrONT’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> CulebrONT
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Table of Contents</a></li>
<li class="toctree-l1"><a class="reference internal" href="#about-this-package">About this package</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#mandatory-installation">Mandatory installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optional-installation">Optional Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#available-data-test">Available data test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#rules-in-culebront">Rules in Culebront</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#from-assembly-to-correction">From assembly to correction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#assembly">Assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="#circularisation">Circularisation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polishing">Polishing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#correction">Correction</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#checking-assembly-quality">Checking assembly quality</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mandatory-quality-tools">Mandatory Quality tools</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional-quality-tools">Optional Quality tools</a></li>
<li class="toctree-l3"><a class="reference internal" href="#report">Report</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#running-culebront">Running CulebrONT</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prepare-config-yaml">Prepare config.yaml</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#providing-data">1. Providing data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chose-assemblers-polisher-and-correctors">2. Chose assemblers, polisher and correctors</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chose-quality-tools">3. Chose quality tools</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parameters-for-some-specific-tools">4. Parameters for some specific tools</a></li>
<li class="toctree-l3"><a class="reference internal" href="#singularity-containers">Singularity containers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#singularity-hub">singularity hub</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#launching-culebront-on-a-single-machine">Launching CulebrONT on a single machine</a></li>
<li class="toctree-l2"><a class="reference internal" href="#launching-culebront-on-hpc-clusters">Launching CulebrONT on HPC clusters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preparing-slurm-cluster-configuration-using-cluster-config-yaml">Preparing Slurm cluster configuration using cluster_config.yaml</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submit-culebront-sh">submit_culebront.sh</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#slurm-wrapper">slurm_wrapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="#profiles">Profiles</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#output-on-culebront">Output on CulebrONT</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">Report</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#useful-notes">Useful notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#thanks">Thanks</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#license">License</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CulebrONT</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Table of Contents</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/README.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <p><img alt="Culebront Logo" src="_images/culebront_logo.png" /></p>
<div class="section" id="table-of-contents">
<h1>Table of Contents<a class="headerlink" href="#table-of-contents" title="Permalink to this headline">¶</a></h1>
<!-- TOC depthFrom:2 depthTo:3 withLinks:1 updateOnSave:1 orderedList:0 --><ul class="simple">
<li><p><a class="reference external" href="#about-this-package">About this package</a></p></li>
<li><p><a class="reference external" href="#installation">Installation</a></p></li>
<li><p><a class="reference external" href="#rules">Rules in CulebrONT</a></p></li>
<li><p><a class="reference external" href="#running-culebront">Running CulebrONT</a></p></li>
<li><p><a class="reference external" href="#Parameters">Parameters</a></p></li>
<li><p><a class="reference external" href="#singularity">Singularity images</a></p></li>
<li><p><a class="reference external" href="#prepclusteryaml">Cluster_config.yaml</a></p></li>
<li><p><a class="reference external" href="#sbatch">Launching CulebrONT on Slurm HPC clusters</a></p></li>
<li><p><a class="reference external" href="#output">Output of CulebrONT</a></p></li>
<li><p><a class="reference external" href="#citation">Citation</a></p></li>
<li><p><a class="reference external" href="#notes">Useful notes</a></p></li>
<li><p><a class="reference external" href="#license">License</a></p></li>
</ul>
<!-- /TOC --><p><a name="about-this-package"></a></p>
</div>
<div class="section" id="about-this-package">
<h1>About this package<a class="headerlink" href="#about-this-package" title="Permalink to this headline">¶</a></h1>
<p>Using data from long reads obtained by Oxford Nanopore Technologies sequencing makes genome assembly easier, in particular to solve repeats and structural variants, in prokaryotic as well as in eukaryotic genomes, resulting in increased contiguity and accuracy.</p>
<p>Bunch of softwares and tools are released or updated every week, and a lot of species see their genome assembled using those.</p>
<p>That’s right.</p>
<p>“<em>But which assembly tool could give the best results for my favorite organism?</em>”</p>
<p><strong>CulebrONT can help you!</strong> CulebrONT is an open-source, scalable, modulable and traceable snakemake pipeline, able to launch multiple assembly tools in parallel and providing help for choosing the best possible assembly between all possibilities.</p>
<p><a name="installation"></a></p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="mandatory-installation">
<h2>Mandatory installation<a class="headerlink" href="#mandatory-installation" title="Permalink to this headline">¶</a></h2>
<p>CulebrONT uses Python &gt;= 3.7 and Snakemake &gt;= 5.10.0. Singularity 3.0.0+ could be of help (see <a class="reference external" href="#singularity">below</a>)</p>
<ul class="simple">
<li><p>Install or update CulebrONT</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone https://github.com/SouthGreenPlatform/CulebrONT_pipeline.git
<span class="nb">cd</span> CulebrONT_pipeline
</pre></div>
</div>
</div>
<div class="section" id="optional-installation">
<h2>Optional Installation<a class="headerlink" href="#optional-installation" title="Permalink to this headline">¶</a></h2>
<p>To obtain a clear and correct report, please add also the following dependencies from R:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">devtools::install_github(&quot;strengejacke/strengejacke&quot;)</span></code></p></li>
<li><p>‘plotly’, ‘dplyr’, ‘optparse’, ‘htmltools’, ‘rmdformats’, ‘magrittr’, ‘yaml’, ‘png’, ‘here’, ‘htmlwidgets’.</p></li>
</ul>
<p><a name="rules"></a></p>
</div>
<div class="section" id="available-data-test">
<h2>Available data test<a class="headerlink" href="#available-data-test" title="Permalink to this headline">¶</a></h2>
<p>A data test <code class="docutils literal notranslate"><span class="pre">Data-Xoo-sub/</span></code> is available on https://itrop.ird.fr/culebront_utilities/. Feel free to download it using <code class="docutils literal notranslate"><span class="pre">wget</span></code> and put it on CulebrONT repertory.</p>
</div>
</div>
<div class="section" id="rules-in-culebront">
<h1>Rules in Culebront<a class="headerlink" href="#rules-in-culebront" title="Permalink to this headline">¶</a></h1>
<p>Assembly, circularisation, polishing and correction steps are included in CulebrONT, and can be activated (or not) according to user’s requests. The most commonly used tools in the community for each step are integrated, as well as various quality control tools. CulebrONT also generates a report compiling information obtained at every step.</p>
<div class="section" id="from-assembly-to-correction">
<h2>From assembly to correction<a class="headerlink" href="#from-assembly-to-correction" title="Permalink to this headline">¶</a></h2>
<p>CulebrONT is really flexible to assembly and circularise (or not) assembled molecules, polish and correct assemblies.</p>
<p>For assemblies, you must launch at least one of assemblers included in CulebrONT and pipe assembly with circularisation, polishing and correction steps as well as with the quality control pipeline.</p>
<p>For circularisation, you can activate/deactivate circularisation steps if needed.
If you are interested on eukaryotic organims, thus circularisation is not necessary use CIRCULAR=False on <em>config.yaml</em> file.</p>
<p>Directed acyclic graphs (DAGs) show the differences between deactivated (left) and activated (right) CIRCULAR step on configuration file.</p>
<p><img alt="global" src="_images/assembly_to_correction.png" /></p>
<div class="section" id="assembly">
<h3>Assembly<a class="headerlink" href="#assembly" title="Permalink to this headline">¶</a></h3>
<p>CulebrONT includes (at the moment) three recent and community-validated assemblers : Flye, Miniasm and Canu. More will be added in the future.</p>
<p>Included tools :</p>
<ul class="simple">
<li><p>Flye version &gt;= 2.6 https://github.com/fenderglass/Flye</p></li>
<li><p>Canu version &gt;= 1.9 https://canu.readthedocs.io/en/latest/quick-start.html</p></li>
<li><p>Miniasm version &gt;= 0.3 https://github.com/lh3/miniasm + Minipolish version &gt;= 0.1.2 https://github.com/rrwick/Minipolish</p></li>
</ul>
</div>
<div class="section" id="circularisation">
<h3>Circularisation<a class="headerlink" href="#circularisation" title="Permalink to this headline">¶</a></h3>
<p>If an assembled molecule is circular, e.g. for bacteria (CIRCULAR=True), this molecule is tagged and will be treated specially in pipeline. We implemented tagging and rotation of circular molecule before each racon polishing step, and we fixing start position on circular genome. This is efficient when multiple genome alignments are envisaged.</p>
<ul class="simple">
<li><p>If Circularisation (CIRCULAR=True) step is choosen, the <em>–plasmids</em> option on Flye is activated.</p></li>
<li><p><em>Ciclator</em> is used to circularise assembly from Canu. Circlator will attempt to identify each circular sequence and output a linearised version from each of them.</p></li>
<li><p>Circularisation for Miniasm is already performed by minipolish.</p></li>
</ul>
<p>Included tools :</p>
<ul class="simple">
<li><p>Circlator version &gt;= 1.5.5 https://github.com/sanger-pathogens/circlator</p></li>
</ul>
</div>
<div class="section" id="polishing">
<h3>Polishing<a class="headerlink" href="#polishing" title="Permalink to this headline">¶</a></h3>
<p>Polishing step is ensured by Racon. Racon corrects raw contigs generated by rapid assembly methods with original ONT reads. Choose how many rounds of Racon you want to perform (constrains from 1 to 9 rounds), and CulebrONT will recursively do it for you. Generally 3 or 4 iterations are the best choices.</p>
<p>Included tools :</p>
<ul class="simple">
<li><p>Racon version &gt;= 1.4.3 https://github.com/isovic/racon</p></li>
</ul>
</div>
<div class="section" id="correction">
<h3>Correction<a class="headerlink" href="#correction" title="Permalink to this headline">¶</a></h3>
<p>Correction can improve the consensus sequence for a draft genome assembly. We include Nanopolish and Medaka on correction steps. With CulebrONT you can now train a Medaka model and use it directly to obtain a consensus from you favorite organism. In addition, Medaka can use a dedicated GPU resource if indicated.</p>
<p>Included tools :s</p>
<ul class="simple">
<li><p>Medaka Medaka-gpu version &gt;= 1.0.3 https://github.com/nanoporetech/medaka</p></li>
<li><p>Nanopolish version &gt;= 0.11.3 https://nanopolish.readthedocs.io/en/latest/index.html#</p></li>
</ul>
</div>
</div>
<div class="section" id="checking-assembly-quality">
<h2>Checking assembly quality<a class="headerlink" href="#checking-assembly-quality" title="Permalink to this headline">¶</a></h2>
<p>A variety of useful tools exist for check high accuracy assembly.</p>
<p><img alt="global" src="_images/quality.png" /></p>
<div class="section" id="mandatory-quality-tools">
<h3>Mandatory Quality tools<a class="headerlink" href="#mandatory-quality-tools" title="Permalink to this headline">¶</a></h3>
<p>In CulebrONT, BUSCO and QUAST are selected by default. BUSCO helps to check if you have a good assembly, by searching the expected single-copy lineage-conserved orthologs in any newly-sequenced genome from an appropriate phylogenetic clade. QUAST is a good starting point to help evaluate the quality of assemblies, providing many helpful contiguity statistics. BUSCO and QUAST statistics are summarized in the CulebrONT final report.</p>
<ul class="simple">
<li><p>BUSCO and QUAST statistics will be calculated by acivaiting the QUALITY (ASSEMBLY, POLISHING, CORRECTION) steps. You must activate at least one over three QUALITY options on <em>config.yaml</em> file.</p></li>
</ul>
<p>Included tools :</p>
<ul class="simple">
<li><p>BUSCO version &gt;= 4.0.5</p></li>
<li><p>QUAST version &gt;= 5.0.2</p></li>
</ul>
</div>
<div class="section" id="optional-quality-tools">
<h3>Optional Quality tools<a class="headerlink" href="#optional-quality-tools" title="Permalink to this headline">¶</a></h3>
<p>CulebrONT checks also the quality of assemblies by using Bloobtools, Assemblytics or KAT, or any combination of these. Weesam can be also used to check read coverage over you assembly (for small genome only). Alignment of several assembles (or from any steps : assembly, polishing, correction) and there comparison is performed using Mauve (for small genome only).</p>
<p>Included tools :</p>
<ul class="simple">
<li><p>Bloobtools version &gt;= 1.1.1</p></li>
<li><p>Assemblytics version &gt;= 1.2</p></li>
<li><p>KAT version &gt;= 2.4.1</p></li>
<li><p>Weesam version &gt; 1.6</p></li>
<li><p>Mauve &gt; 2.4.0.snapshot_2015_02_13</p></li>
</ul>
</div>
<div class="section" id="report">
<h3>Report<a class="headerlink" href="#report" title="Permalink to this headline">¶</a></h3>
<p>CulebrONT generates a report presenting the summary statistics from different steps of the pipeline.</p>
<p>This report has different dependencies to be preinstalled, as recomanded in the <a class="reference external" href="#installation">installation step</a></p>
<p>Don’t worry if you do not have access to all these depedencies !! The whole packages used in the report are available in  the <em>R.def</em> Singularity image, available at the Singularity hub or build from the <a class="reference external" href="https://github.com/SouthGreenPlatform/CulebrONT_pipeline/tree/master/Containers">CulebrONT Containers repository</a> (only if you have sudo super cowpowers).</p>
<p><a name="running-culebront"></a></p>
</div>
</div>
</div>
<div class="section" id="running-culebront">
<h1>Running CulebrONT<a class="headerlink" href="#running-culebront" title="Permalink to this headline">¶</a></h1>
<p><strong>For IRD iTrop or IFB HPC resources users ONLY</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># On i-Trop HPC</span>
module load system/singularity/3.3.0
module load system/python/3.7.2

<span class="c1"># On IFB HPC</span>
module load singularity
module load python/3.7
module load graphviz/2.40.
</pre></div>
</div>
<div class="section" id="prepare-config-yaml">
<h2>Prepare config.yaml<a class="headerlink" href="#prepare-config-yaml" title="Permalink to this headline">¶</a></h2>
<p>To run the pipeline you have to provide the data path and activate/deactivate options in every step from config.yaml file.</p>
<div class="section" id="providing-data">
<h3>1. Providing data<a class="headerlink" href="#providing-data" title="Permalink to this headline">¶</a></h3>
<p>First, indicate the data path in the configuration file</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DATA</span><span class="p">:</span>
    <span class="nt">FASTQ</span><span class="p">:</span> <span class="s">&#39;/path/to/fastq/directory/&#39;</span>
    <span class="nt">REF</span><span class="p">:</span> <span class="s">&#39;/path/to/referencefile.fasta&#39;</span>
    <span class="nt">GENOME_SIZE</span><span class="p">:</span> <span class="s">&#39;1m&#39;</span>
    <span class="nt">FAST5</span><span class="p">:</span> <span class="s">&#39;/path/to/fast5/directory/&#39;</span>
    <span class="nt">ILLUMINA</span><span class="p">:</span> <span class="s">&#39;/path/to/illumina/directory/&#39;</span>
    <span class="nt">OUTPUT</span><span class="p">:</span> <span class="s">&#39;/path/to/output/directory/&#39;</span>
    <span class="nt">CIRCULAR </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True/False</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>FASTQ</strong>: CulebrONT takes as input a FASTQ directory. Every FASTQ file should contain the whole set of reads to be assembled (meaning that multiple runs must be merged in a single FASTQ file), as each FASTQ file found in this repertory will be assembled independently. FASTQ files can be compressed or not (gzipped). Naming convention accepted by CulebrONT are: <em>NAME.fastq.gz</em> or <em>NAME.fq.gz</em> or <em>NAME.fastq</em> or <em>NAME.fq</em>.</p></li>
<li><p><strong>REF</strong>: Only one REFERENCE genome file will be used by CulebrONT. This REFERENCE will be used for QC steps (QUAST and MAUVE).</p></li>
<li><p><strong>GENOME_SIZE</strong> : Estimated genome size (m,g,k) of the assembly.</p></li>
<li><p><strong>FAST5</strong>: Nanopolish needs FAST5 files to training steps. Please give the path of FAST5 repertory in the <em>FAST5</em> DATA parameter. Inside this directory, a subdirectory with the exact same name as the corresponding FASTQ (before the <em>.fastq.gz</em>) is requested. For instance, if in the <em>FASTQ</em> directory we have <em>run1.fastq.gz</em> and <em>run2.fastq.gz</em>, CulebrONT is expecting the <em>run1/</em> and <em>run2/</em> subdirectories in the FAST5 main directory.</p></li>
<li><p><strong>ILLUMINA</strong> : indicate the path to the directory with <em>Illumina</em> sequence data (in fastq or fastq.gz format) to perform KAT quality. Use preferentially paired-end data.</p></li>
<li><p><strong>OUTPUT</strong>: output <em>path</em> directory.</p></li>
<li><p><strong>CIRCULAR</strong> : Indicate <em>True</em> or <em>False</em> to activate/deactivate circularisation steps (only to procaryote).</p></li>
</ul>
</div>
<div class="section" id="chose-assemblers-polisher-and-correctors">
<h3>2. Chose assemblers, polisher and correctors<a class="headerlink" href="#chose-assemblers-polisher-and-correctors" title="Permalink to this headline">¶</a></h3>
<p>Activate/deactivate assemblers and correctors as you wish. By default, Racon is launched as POLISH tool after each activated assembly step. You <strong>must</strong> activate at least one assembler and one corrector.</p>
<p>Example:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">ASSEMBLY</span><span class="p">:</span>
    <span class="nt">CANU </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>
    <span class="nt">FLYE </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">MINIASM </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="nt">CORRECTION</span><span class="p">:</span>
    <span class="nt">NANOPOLISH </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">MEDAKA </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>
</pre></div>
</div>
</div>
<div class="section" id="chose-quality-tools">
<h3>3. Chose quality tools<a class="headerlink" href="#chose-quality-tools" title="Permalink to this headline">¶</a></h3>
<p>Activate/deactivate quality tools as you wish. By default, BUSCO AND QUAST are launched if ‘True’ in QUALITY (ASSEMBLY OR POLISHING OR CORRECTION OR both) steps.</p>
<p>You <strong>must</strong> to activate at least one QUALITY step.</p>
<p>Example:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">QUALITY</span><span class="p">:</span>
    <span class="nt">ASSEMBLY </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">POLISHING </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">CORRECTION </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<p>Others quality tools are launched only on the <em>final assemblies</em> (BLOBTOOLS, ASSEMBLYTICS, WEESAM and KAT).</p>
<p>KAT quality tool can be activate but Illumina reads are mandatory in this case.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">#### Others quality tools</span>
    <span class="nt">WEESAM</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">BLOBTOOLS</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">ASSEMBLYTICS</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">KAT</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<p>Alignment of various assemblies <strong>for small genomes (&lt;10-20Mbp)</strong> is also possible by using Mauve. Mauve will compared each state of the assembly (Raw, Polished and Corrected) for each assembler used.</p>
<p>A <em>Fixstart</em> step is possible before Mauve MSA to improve alignment on circular molecules.</p>
<ul class="simple">
<li><p>Fixstart will be deactivated if CIRCULAR is False</p></li>
<li><p>Only activate MAUVE if you have more than one sample and more than one quality step.</p></li>
</ul>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">MSA</span><span class="p">:</span>
    <span class="nt">FIXSTART</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
    <span class="nt">MAUVE</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<p><a name="parameters"></a></p>
</div>
<div class="section" id="parameters-for-some-specific-tools">
<h3>4. Parameters for some specific tools<a class="headerlink" href="#parameters-for-some-specific-tools" title="Permalink to this headline">¶</a></h3>
<p>Specifically to Racon:</p>
<ul class="simple">
<li><p>Racon can be launch recursively from 1 to 9 rounds. 3 or 4 are recommanded.</p></li>
</ul>
<p>Specifically to Medaka :</p>
<ul class="simple">
<li><p>If ‘MEDAKA_TRAIN_WITH_REF’ is activated, Medaka launchs training using the reference found in ‘DATA/REF’ param. Medaka does not take into account other medaka model parameters.</p></li>
<li><p>If ‘MEDAKA_TRAIN_WITH_REF’ is deactivated, Medaka does not launch training but uses instead the model provided in ‘MEDAKA_MODEL_PATH’. If ‘MEDAKA_MODEL_PATH’ is empty, this param is not used and the default model for <em>E.coli</em> is used.</p></li>
</ul>
<p>Standard parameters used:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">############ PARAMS ################</span>
<span class="nt">params</span><span class="p">:</span>
    <span class="nt">MINIMAP2</span><span class="p">:</span>
        <span class="nt">PRESET_OPTION</span><span class="p">:</span> <span class="s">&#39;map-pb&#39;</span> <span class="c1"># -x minimap2 preset option is map-pb by default (map-pb, map-ont etc)</span>
    <span class="nt">CANU</span><span class="p">:</span>
        <span class="nt">MAX_MEMORY</span><span class="p">:</span> <span class="s">&#39;15G&#39;</span>
        <span class="nt">OPTIONS</span><span class="p">:</span> <span class="s">&#39;-fast&#39;</span>
    <span class="nt">CIRCLATOR</span><span class="p">:</span>
        <span class="nt">OPTIONS</span><span class="p">:</span> <span class="s">&#39;&#39;</span>
    <span class="nt">RACON</span><span class="p">:</span>
        <span class="nt">RACON_ROUNDS</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2</span> <span class="c1">#1 to 9</span>
    <span class="nt">NANOPOLISH</span><span class="p">:</span>
        <span class="c1"># segment length to split assembly and correct it  default=50000</span>
        <span class="nt">NANOPOLISH_SEGMENT_LEN</span><span class="p">:</span> <span class="s">&#39;50000&#39;</span>
        <span class="c1"># overlap length between segments  default=200</span>
        <span class="nt">NANOPOLISH_OVERLAP_LEN</span><span class="p">:</span> <span class="s">&#39;200&#39;</span>
        <span class="nt">OPTIONS</span><span class="p">:</span> <span class="s">&#39;&#39;</span>
    <span class="nt">MEDAKA</span><span class="p">:</span>
        <span class="c1"># if &#39;MEDAKA_TRAIN_WITH_REF&#39; is True, Medaka launchs training using the reference found in DATA REF param. Medaka does not take in count other Medaka model parameters below.</span>
        <span class="nt">MEDAKA_TRAIN_WITH_REF</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>
        <span class="nt">MEDAKA_MODEL_PATH</span><span class="p">:</span> <span class="s">&#39;medakamodel/r941_min_high_g303_model.hdf5&#39;</span> <span class="c1"># if empty this param is not used.</span>
    <span class="nt">BUSCO</span><span class="p">:</span>
        <span class="nt">DATABASE </span><span class="p">:</span> <span class="s">&#39;Data-Xoo-sub/bacteria_odb10&#39;</span>
        <span class="nt">MODEL </span><span class="p">:</span> <span class="s">&#39;genome&#39;</span>
<span class="c1">#        &#39;SP&#39; : &#39;caenorhabditis&#39;</span>
        <span class="nt">SP </span><span class="p">:</span> <span class="s">&#39;&#39;</span>
    <span class="nt">QUAST</span><span class="p">:</span>
        <span class="nt">REF</span><span class="p">:</span> <span class="s">&#39;Data-Xoo-sub/ref/BAI3_Sanger.fsa&#39;</span>
        <span class="nt">GFF</span><span class="p">:</span> <span class="s">&#39;&#39;</span>
        <span class="nt">GENOME_SIZE_PB</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">48000000</span>
        <span class="c1">#GENOME_SIZE_PB: 1000000</span>
        <span class="nt">OPTIONS </span><span class="p">:</span> <span class="s">&#39;&#39;</span>
    <span class="nt">DIAMOND</span><span class="p">:</span>
        <span class="nt">DATABASE</span><span class="p">:</span> <span class="s">&#39;Data-Xoo-sub/testBacteria.dmnd&#39;</span>
    <span class="nt">MUMMER</span><span class="p">:</span>
<span class="c1">#         -l default 20</span>
        <span class="nt">MINMATCH </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="c1">#         -c default 65</span>
        <span class="nt">MINCLUSTER</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">500</span>
    <span class="nt">ASSEMBLYTICS</span><span class="p">:</span>
        <span class="nt">UNIQUE_ANCHOR_LEN</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">10000</span>
        <span class="nt">MIN_VARIANT_SIZE</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">50</span>
        <span class="nt">MAX_VARIANT_SIZE</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">10000</span>
</pre></div>
</div>
<p><a name="singularity"></a></p>
</div>
<div class="section" id="singularity-containers">
<h3>Singularity containers<a class="headerlink" href="#singularity-containers" title="Permalink to this headline">¶</a></h3>
<p>To use Singularity containers, provide to CulebrONT the already build Singularity containers path on your computer or cluster.</p>
<p>As an example, here are singularity images found  on the i-Trop HPC from the SouthGreen platform.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster with scratch temporary directory</span>
<span class="l l-Scalar l-Scalar-Plain">SCRATCH = False</span>

<span class="l l-Scalar l-Scalar-Plain">## @ITROP PATH</span>
<span class="l l-Scalar l-Scalar-Plain">tools</span><span class="p p-Indicator">:</span>
<span class="c1">## ASSEMBLERS:</span>
    <span class="nt">CANU_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/canu-1.9.simg&#39;</span>
    <span class="nt">FLYE_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/flye-2.6.simg&#39;</span>
    <span class="nt">MINIASM_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/miniasm-0.3.simg&#39;</span>
    <span class="nt">MINIPOLISH_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/minipolish-0.1.2.simg&#39;</span>
<span class="c1">## CIRCULARISATION</span>
    <span class="nt">CIRCLATOR_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/circlator-1.5.5.simg&#39;</span>
<span class="c1">## POLISHERS:</span>
    <span class="nt">RACON_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/racon-1.4.3.simg&#39;</span>
    <span class="nt">NANOPOLISH_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg&#39;</span>
<span class="c1">## CORRECTION</span>
    <span class="nt">MEDAKA_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/medaka_conda-gpu-1.0.3.simg&#39;</span>
<span class="c1">## QUALITY</span>
    <span class="nt">BUSCO_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/busco-4.0.5.simg&#39;</span>
    <span class="nt">QUAST_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/quast-5.0.2.simg&#39;</span>
    <span class="nt">WEESAM_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/weesam.simg&#39;</span>
    <span class="nt">BLOBTOOLS_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/bloobtools-v1.1.1.simg&#39;</span>
    <span class="nt">MINIMAP2_SIMG</span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg&#39;</span>
    <span class="nt">DIAMOND_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/diamond-0.9.30.simg&#39;</span>
    <span class="nt">MUMMER_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/mummer-4beta.simg&#39;</span>
    <span class="nt">ASSEMBLYTICS_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/assemblytics-1.2.simg&#39;</span>
    <span class="nt">SAMTOOLS_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/nanopolish-0.11.3.simg&#39;</span>
    <span class="nt">KAT_SIMG </span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/kat-2.4.2.simg&#39;</span>
    <span class="nt">MINICONDA_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://vibaotram/singularity-container:cpu-guppy3.4-conda-api&#39;</span>
    <span class="nt">R_SIMG</span><span class="p">:</span> <span class="s">&#39;/data3/projects/containers/CULEBRONT/R.simg&#39;</span>
</pre></div>
</div>
<p>Available recipes from containers are available in the <em>Containers</em> folder, as well as on the main <a class="reference external" href="https://github.com/SouthGreenPlatform/CulebrONT.git">CulebrONT repository</a>. Feel free to build them on your own computer (or cluster); be careful, you need root rights to do it.</p>
<p>Built singularity images are also available on https://itrop.ird.fr/culebront_utilities/. Feel free to download it using <code class="docutils literal notranslate"><span class="pre">wget</span></code> for example.</p>
<div class="section" id="singularity-hub">
<h4>singularity hub<a class="headerlink" href="#singularity-hub" title="Permalink to this headline">¶</a></h4>
<p>If you want to recover singularity images from the Singularity Hub and build them, please use these paths :</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster with scratch temporal repertory</span>
<span class="nt">SCRATCH </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>

<span class="nt">tools</span><span class="p">:</span>
<span class="c1">###### ASSEMBLERS:</span>
      <span class="nt">CANU_SIMG</span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:canu-1.9.def&#39;</span>
      <span class="nt">FLYE_SIMG</span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:flye-2.6.def&#39;</span>
      <span class="nt">MINIASM_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:miniasm-0.3.def&#39;</span>
      <span class="nt">MINIPOLISH_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:minipolish-0.1.2.def&#39;</span>
<span class="c1">###### CIRCULARISATION</span>
      <span class="nt">CIRCLATOR_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:circlator-1.5.5.def&#39;</span>
<span class="c1">###### POLISHERS:</span>
      <span class="nt">RACON_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:racon-1.4.3.def&#39;</span>
<span class="c1">###### CORRECTION</span>
      <span class="nt">NANOPOLISH_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:nanopolish-0.11.3.def&#39;</span>
      <span class="nt">MEDAKA_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:medaka_conda-gpu-1.0.3.simg&#39;</span>
<span class="c1">###### QUALITY</span>
      <span class="nt">BUSCO_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:busco-4.def&#39;</span>
      <span class="nt">QUAST_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">WEESAM_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">BLOBTOOLS_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">MINIMAP2_SIMG</span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:nanopolish-0.11.3.simg&#39;</span>
      <span class="nt">DIAMOND_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">MUMMER_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:mummer-4beta.def&#39;</span>
      <span class="nt">ASSEMBLYTICS_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">SAMTOOLS_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:nanopolish-0.11.3.simg&#39;</span>
      <span class="nt">KAT_SIMG </span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:quality.def&#39;</span>
      <span class="nt">MINICONDA_SIMG</span><span class="p">:</span> <span class="s">&#39;shub://vibaotram/singularity-container:cpu-guppy3.4-conda-api&#39;</span>
      <span class="nt">R_SIMG</span><span class="p">:</span> <span class="s">&#39;shub://SouthGreenPlatform/CulebrONT_pipeline:r.def&#39;</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="launching-culebront-on-a-single-machine">
<h2>Launching CulebrONT on a single machine<a class="headerlink" href="#launching-culebront-on-a-single-machine" title="Permalink to this headline">¶</a></h2>
<p><a name="Local"></a></p>
<p>To launch CulebrONT on a single machine, you should use the parameters <code class="docutils literal notranslate"><span class="pre">--use-singularity</span></code> and <code class="docutils literal notranslate"><span class="pre">--use-conda</span></code>.</p>
<p>See the example below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">-</span><span class="n">s</span> <span class="n">Snakefile</span> <span class="o">--</span><span class="n">configfile</span> <span class="n">config</span><span class="o">.</span><span class="n">yaml</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">singularity</span>  <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="o">--</span><span class="n">latency</span><span class="o">-</span><span class="n">wait</span> <span class="mi">120</span>
</pre></div>
</div>
<p><a name="sbatch"></a></p>
</div>
<div class="section" id="launching-culebront-on-hpc-clusters">
<h2>Launching CulebrONT on HPC clusters<a class="headerlink" href="#launching-culebront-on-hpc-clusters" title="Permalink to this headline">¶</a></h2>
<p><a name="prepclusteryaml"></a></p>
<div class="section" id="preparing-slurm-cluster-configuration-using-cluster-config-yaml">
<h3>Preparing Slurm cluster configuration using cluster_config.yaml<a class="headerlink" href="#preparing-slurm-cluster-configuration-using-cluster-config-yaml" title="Permalink to this headline">¶</a></h3>
<p>On <code class="docutils literal notranslate"><span class="pre">cluster_config.yaml</span></code>, you can add partition, memory and threads to be used by default for each rule. If more memory or threads are requested, please adapt the content of this file before running on a cluster for every rule. For instance give more memory to Canu and Medaka.</p>
<p>Here is a example of the configuration file we used on the i-Trop HPC.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">__default__</span><span class="p">:</span>
    <span class="nt">cpus-per-task </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">4</span>
    <span class="nt">ntasks </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">1</span>
    <span class="nt">mem-per-cpu </span><span class="p">:</span> <span class="s">&#39;2&#39;</span>
    <span class="nt">partition </span><span class="p">:</span> <span class="s">&quot;normal&quot;</span>
    <span class="nt">output </span><span class="p">:</span> <span class="s">&#39;logs/stdout/{rule}/{wildcards}&#39;</span>
    <span class="nt">error </span><span class="p">:</span> <span class="s">&#39;logs/error/{rule}/{wildcards}&#39;</span>
<span class="c1">#</span>
<span class="c1">#run_nanopolish :</span>
<span class="c1">#    cpus-per-task : 8</span>
<span class="c1">#    mem-per-cpu : &#39;4&#39;</span>
<span class="c1">#    partition : &quot;long&quot;</span>
<span class="c1">#</span>
<span class="nt">run_canu</span><span class="p">:</span>
    <span class="nt">cpus-per-task </span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">8</span>
    <span class="nt">mem-per-cpu </span><span class="p">:</span> <span class="s">&#39;4&#39;</span>
    <span class="nt">partition </span><span class="p">:</span> <span class="s">&quot;long&quot;</span>
</pre></div>
</div>
<div class="section" id="submit-culebront-sh">
<h4>submit_culebront.sh<a class="headerlink" href="#submit-culebront-sh" title="Permalink to this headline">¶</a></h4>
<p>This is a typical launcher for using CulebrONT on a Slurm cluster. You have to adapt it for the configuration of your favorite cluster.
Please adapt this script also if you want to use wrappers or profiles.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1">#SBATCH --job-name culebrONT</span>
<span class="c1">#SBATCH --output slurm-%x_%j.log</span>
<span class="c1">#SBATCH --error slurm-%x_%j.log</span>

<span class="n">module</span> <span class="n">load</span> <span class="n">system</span><span class="o">/</span><span class="n">singularity</span><span class="o">/</span><span class="mf">3.3</span><span class="o">.</span><span class="mi">0</span>
<span class="n">module</span> <span class="n">load</span> <span class="n">system</span><span class="o">/</span><span class="n">python</span><span class="o">/</span><span class="mf">3.7</span><span class="o">.</span><span class="mi">2</span>

<span class="n">snakemake</span> <span class="o">--</span><span class="n">unlock</span>

<span class="c1"># SLURM JOBS WITH USING WRAPPER</span>
<span class="n">snakemake</span> <span class="o">--</span><span class="n">nolock</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">singularity</span> <span class="o">--</span><span class="n">cores</span> <span class="o">-</span><span class="n">p</span> <span class="o">--</span><span class="n">verbose</span> <span class="o">-</span><span class="n">s</span> <span class="n">Snakefile</span> \
<span class="o">--</span><span class="n">latency</span><span class="o">-</span><span class="n">wait</span> <span class="mi">60</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">going</span> <span class="o">--</span><span class="n">restart</span><span class="o">-</span><span class="n">times</span> <span class="mi">1</span> <span class="o">--</span><span class="n">rerun</span><span class="o">-</span><span class="n">incomplete</span>  \
<span class="o">--</span><span class="n">configfile</span> <span class="n">config</span><span class="o">-</span><span class="n">itrop</span><span class="o">.</span><span class="n">yaml</span> \
<span class="o">--</span><span class="n">cluster</span> <span class="s2">&quot;python3 slurm_wrapper.py&quot;</span> \
<span class="o">--</span><span class="n">cluster</span><span class="o">-</span><span class="n">config</span> <span class="n">cluster_config</span><span class="o">.</span><span class="n">yaml</span> \
<span class="o">--</span><span class="n">cluster</span><span class="o">-</span><span class="n">status</span> <span class="s2">&quot;python3 slurm_status.py&quot;</span>

<span class="c1"># USING PROFILES</span>
<span class="c1">#snakemake --nolock --use-singularity --use-conda --cores -p --verbose -s Snakefile --configfile config-itrop.yaml \</span>
<span class="c1">#--latency-wait 60 --keep-going --restart-times 1 --rerun-incomplete --cluster-config cluster_config.yaml --profile slurm-culebrONT </span>
</pre></div>
</div>
<p>This launcher can be submitted to the Slurm queue typing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sbatch</span> <span class="n">submit_culebront</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p><em>Important</em> : Do not forget to adapt submit_culebront.sh if you want to use wrappers or profile!!</p>
</div>
</div>
<div class="section" id="slurm-wrapper">
<h3>slurm_wrapper<a class="headerlink" href="#slurm-wrapper" title="Permalink to this headline">¶</a></h3>
<p>A slurm_wrapper.py script is available on CulebrONT projet to manage ressources from your cluster configuration (taken from cluster_config.yaml file).
This is the easier way to know what is running on cluster and to adapt ressources for every job. Take care, this cluster_config.yaml file is becomming obsolete on next Snakemake versions.</p>
</div>
<div class="section" id="profiles">
<h3>Profiles<a class="headerlink" href="#profiles" title="Permalink to this headline">¶</a></h3>
<p>Optionally is possible to use Profiles in order to run CulebrONT on HPC cluster. Please follow the <a class="reference external" href="https://github.com/Snakemake-Profiles/">recommandations found on the SnakeMake profile github</a>.</p>
<p>Here is an example of how to profile a Slurm scheduler using <a class="reference external" href="https://github.com/Snakemake-Profiles/slurm">those recommandations</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p ~/.config/snakemake
$ cd ~/.config/snakemake
$ pip install --user cookiecutter
$ cookiecutter https://github.com/Snakemake-Profiles/slurm.git
# Answer to questions :
profile_name [slurm]: slurm-culebrONT
sbatch_defaults []:
cluster_config []: cluster_config.yaml
Select advanced_argument_conversion:
1 - no
2 - yes
Choose from 1, 2 [1]: 2

$ cp /shared/home/$USER/CulebrONT_pipeline/cluster_config.yaml .
# edit slurm-submit.py and upgrade this dictionary with your Slurm options

RESOURCE_MAPPING = {
	&quot;cpus-per-task&quot;: (&quot;cpus-per-task&quot;, &quot;cpu&quot;),
	&quot;partition&quot;: (&quot;queue&quot;, &quot;partition&quot;),
	&quot;time&quot;: (&quot;time&quot;, &quot;runtime&quot;, &quot;walltime&quot;),
	&quot;mem&quot;: (&quot;mem&quot;, &quot;mem_mb&quot;, &quot;ram&quot;, &quot;memory&quot;),
	&quot;mem-per-cpu&quot;: (&quot;mem-per-cpu&quot;, &quot;mem_per_cpu&quot;, &quot;mem_per_thread&quot;),
	&quot;nodes&quot;: (&quot;nodes&quot;, &quot;nnodes&quot;),
	&quot;error&quot;: (&quot;logerror&quot;, &quot;e&quot;),
	&quot;output&quot;: (&quot;logout&quot;, &quot;o&quot;)}
</pre></div>
</div>
<p>SLURM Profile <em>slurm-culebrONT</em> is now created on : <code class="docutils literal notranslate"><span class="pre">/shared/home/$USER/.config/snakemake/slurm-culebrONT</span></code> repertory</p>
<p><a name="output"></a></p>
</div>
</div>
</div>
<div class="section" id="output-on-culebront">
<h1>Output on CulebrONT<a class="headerlink" href="#output-on-culebront" title="Permalink to this headline">¶</a></h1>
<p>The architecture of CulebrONT output is designed as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>output_example_circ/
├── SAMPLE-1
│   ├── CANU
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   ├── FLYE
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   ├── MAUVE_ALIGN
│   ├── MINIASM
│   │   ├── ASSEMBLER
│   │   ├── CORRECTION
│   │   ├── MSA
│   │   ├── POLISHING
│   │   └── QUALITY
│   └── QUAST
│       ├── data
│       └── quast_results
├── LOGS
...
└── REPORT
    └── SAMPLE-1
</pre></div>
</div>
<p>The same Architecture per sample (fastq = SAMPLE-1 in example) is followed for LOG files.</p>
<div class="section" id="id1">
<h2>Report<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>CulebrONT generates a beautiful report containing, foreach fastq found on input directory, a summary of interesting statistics such as busco, quast and others ones that you will discover!</p>
<p><img alt="Culebront Logo" src="_images/report.png" /></p>
<p><em>Important</em>: To visualise the report created by CulebrONT, transfer the whole <em>REPORT</em> directory on your local machine before opening the <em>report.html</em> file with a navigator.</p>
<p><a name="citation"></a></p>
</div>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p>&#64;Authors:</p>
<p>Aurore Comte (IRD) and Julie Orjuela (IRD) developped CulebrONT.</p>
<p>Sebastien Ravel (CIRAD), Florian Charriat (CIRAD) helped on SnakeMake bugs and tricks.</p>
<p>François Sabot (IRD) was involved in coverage tools implementation.</p>
<p>Ndomassi Tando (IRD) helped on Singularity containers creation.</p>
<p>Sebastien Cunnac (IRD) and Tram Vi (IRD) tested on real data and helped for circularisation improvements.</p>
<p><a name="notes"></a></p>
</div>
<div class="section" id="useful-notes">
<h1>Useful notes<a class="headerlink" href="#useful-notes" title="Permalink to this headline">¶</a></h1>
<p>Before launching CulebrONT, you could base-calling of arbitrarily multiplexed libraries across several Minion runs with sequencing quality control and gather the output files by genome for subsequent steps. For that use https://github.com/vibaotram/baseDmux.</p>
<div class="section" id="thanks">
<h2>Thanks<a class="headerlink" href="#thanks" title="Permalink to this headline">¶</a></h2>
<p>The authors acknowledge the IRD i-Trop HPC (South Green Platform) at IRD Montpellier for providing HPC resources that have contributed to this work. https://bioinfo.ird.fr/ - http://www.southgreen.fr</p>
<p>Thanks to Yann Delorme for this beautiful logo https://nimarell.github.io/resume</p>
<p><a name="licence"></a></p>
</div>
</div>
<div class="section" id="license">
<h1>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>Licencied under CeCill-C (http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html) and GPLv3
Intellectual property belongs to IRD and authors.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to culebrONT’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2019-2020, Julie Orjuela (IRD), Aurore Comte (IRD), Sebastien Ravel (CIRAD), Florian Charriat (CIRAD)

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>